### PR TITLE
Added shadow delete feature to assets, asset_tags, and contact_events

### PR DESCRIPTION
**Pull request by xunchuan-liu**
# Changes
- 3 migrations to add new boolean field to the affected tables
- Changed delete implementation to only toggle the boolean value
- Updated models, routes, and tests to reflect changes
- Added function to delete all asset tags associated with an asset

### CODE REVIEW

File: migrations/2021-03-10-220926_add_deleted_to_assets/down.sql [hunk 1]
```diff
@@ -0,0 +1,4 @@
+-- This file should undo anything in `up.sql`
```
Comment (xunchuan-liu) [inline migrations/2021-03-10-220926_add_deleted_to_assets/down.sql:1]: The migrations as follows for altering the columns

File: src/asset_tags/model.rs [hunk 2]
```diff
@@ -1,22 +1,25 @@
+use crate::assets::Asset;
```
Comment (xunchuan-liu) [inline src/asset_tags/model.rs:1]: Changes below for all the models and routes after adding new field

File: src/asset_tags/model.rs [hunk 3]
```diff
@@ -25,20 +28,38 @@ pub struct MaybeAssetTag {
...
 
 impl AssetTag {
     pub fn find_all() -> Result<Vec<Self>, CustomError> {
+        let conn = db::connection()?;
+        let asset_tags = asset_tags::table
+            .filter(asset_tags::deleted.eq(false))
+            .load::<AssetTag>(&conn)?;
+        Ok(asset_tags)
+    }
+
```
Comment (xunchuan-liu) [inline src/asset_tags/model.rs:43]: New filter to find all of them, including deleted assets

File: src/asset_tags/model.rs [hunk 4]
```diff
@@ -25,20 +28,38 @@ pub struct MaybeAssetTag {
...
+    pub fn find_with_deleted() -> Result<Vec<Self>, CustomError> {
         let conn = db::connection()?;
         let asset_tags = asset_tags::table.load::<AssetTag>(&conn)?;
         Ok(asset_tags)
     }
```
Comment (xunchuan-liu) [inline src/asset_tags/model.rs:49]: New filter to find only deleted assets

File: src/asset_tags/model.rs [hunk 5]
```diff
@@ -71,20 +94,29 @@ impl AssetTag {
         let conn = db::connection()?;
         let asset_tag = diesel::update(asset_tags::table)
             .filter(asset_tags::id.eq(id))
+            .filter(asset_tags::deleted.eq(false))
             .set(asset_tag)
             .get_result(&conn)?;
         Ok(asset_tag)
     }
```
Comment (xunchuan-liu) [inline src/asset_tags/model.rs:102]: Rewrite delete as a single field update

File: src/asset_tags/model.rs [hunk 6]
```diff
@@ -71,20 +94,29 @@ impl AssetTag {
...
         let conn = db::connection()?;
-        let res = diesel::delete(asset_tags::table.filter(asset_tags::id.eq(id))).execute(&conn)?;
-        Ok(res)
+        let asset_tags = diesel::update(asset_tags::table)
+            .filter(asset_tags::id.eq(id))
+            .filter(asset_tags::deleted.eq(false))
+            .set(asset_tags::deleted.eq(true))
+            .get_result(&conn)?;
+        Ok(asset_tags)
     }
```
Comment (xunchuan-liu) [inline src/asset_tags/model.rs:112]: Delete all asset tags by asset id

File: src/schema.rs [hunk 7]
```diff
@@ -104,7 +107,6 @@ table! {
 }
 
 joinable!(alerts -> users (user_id));
-joinable!(assets -> asset_tags (asset_tag_id));
```
Comment (xunchuan-liu) [inline src/schema.rs:107]: Looks like every time I let the `schema.rs` get autogenerated, it wants to take out this joinable line. I would need to add it back in manually every time. Is there another way around this?

Comment (trueb2) [reply src/schema.rs:107]: I don't see a fix for this online. There is a dependency update we could upgrade to. Also you could try upgrading your diesel_cli

**Review by trueb2 — APPROVED** [review]
This is exactly what I was looking for 👍🏻

**Pull request merged**